File: app/src/main/java/com/stupidtree/hichat/ui/relation/RelationActivity.java
package com.stupidtree.hichat.ui.relation;

import android.os.Bundle;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.Toolbar;

import com.stupidtree.hichat.R;
import com.stupidtree.hichat.data.model.UserRelation;
import com.stupidtree.hichat.ui.base.BaseActivity;
import com.stupidtree.hichat.ui.base.DataState;
import com.stupidtree.hichat.ui.widgets.PopUpEditText;

import butterknife.BindView;

public class RelationActivity extends BaseActivity<RelationViewModel> {

    @BindView(R.id.remark_layout)
    ViewGroup remarkLayout;

    @BindView(R.id.remark)
    TextView remarkText;


    @BindView(R.id.toolbar)
    Toolbar toolbar;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setWindowParams(true, true, false);
        setToolbarActionBack(toolbar);
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_relation;
    }

    @Override
    protected Class<RelationViewModel> getViewModelClass() {
        return RelationViewModel.class;
    }

    @Override
    protected void initViews() {
        remarkLayout.setOnClickListener(view -> {
            DataState<UserRelation> up = viewModel.getRelationData().getValue();
            if (up != null && up.getState() == DataState.STATE.SUCCESS) {
                new PopUpEditText()
                        .setTitle(R.string.prompt_set_remark)
                        .setText(up.getData().getRemark())
                        .setOnConfirmListener(text -> {
                            //控制viewModel发起更改昵称请求
                            viewModel.startChangeRemark(text);
                        })
                        .show(getSupportFragmentManager(), "edit");
            }

        });
        viewModel.getChangeRemarkResult().observe(this, stringDataState -> {
            if (stringDataState.getState() == DataState.STATE.SUCCESS) {
                Toast.makeText(getThis(), R.string.avatar_change_success, Toast.LENGTH_SHORT).show();
                viewModel.startFetchRelationData(getIntent().getStringExtra("friendId"));
            } else {
                Toast.makeText(getApplicationContext(), "失败", Toast.LENGTH_SHORT).show();
            }
        });

        viewModel.getRelationData().observe(this, friendContactDataState -> {
            if (friendContactDataState.getState() == DataState.STATE.SUCCESS) {
                remarkText.setText(friendContactDataState.getData().getRemark());
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (getIntent().hasExtra("friendId")) {
            viewModel.startFetchRelationData(getIntent().getStringExtra("friendId"));
        }
    }
}

File: app/src/main/java/com/stupidtree/cloudliter/ui/relation/RelationActivity.kt
package com.stupidtree.cloudliter.ui.relation

import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.Observer
import butterknife.BindView
import com.stupidtree.cloudliter.R
import com.stupidtree.cloudliter.ui.base.BaseActivity
import com.stupidtree.cloudliter.ui.base.DataState
import com.stupidtree.cloudliter.ui.widgets.PopUpEditText

class RelationActivity : BaseActivity<RelationViewModel>() {
    @JvmField
    @BindView(R.id.remark_layout)
    var remarkLayout: ViewGroup? = null

    @JvmField
    @BindView(R.id.remark)
    var remarkText: TextView? = null

    @JvmField
    @BindView(R.id.toolbar)
    var toolbar: Toolbar? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setWindowParams(true, true, false)
        setToolbarActionBack(toolbar!!)
    }

    override fun getLayoutId(): Int {
        return R.layout.activity_relation
    }

    override fun getViewModelClass(): Class<RelationViewModel> {
        return RelationViewModel::class.java
    }

    override fun initViews() {
        remarkLayout!!.setOnClickListener { view: View? ->
            val up = viewModel!!.relationData!!.value
            if (up != null && up.state === DataState.STATE.SUCCESS) {
                PopUpEditText()
                        .setTitle(R.string.prompt_set_remark)
                        .setText(up.data!!.remark)
                        .setOnConfirmListener(object :PopUpEditText.OnConfirmListener{
                            override fun OnConfirm(text: String) {
                                //控制viewModel发起更改昵称请求
                                viewModel!!.startChangeRemark(text)
                            }
                        })
                        .show(supportFragmentManager, "edit")
            }
        }
        viewModel!!.changeRemarkResult!!.observe(this, Observer { stringDataState: DataState<String?>? ->
            if (stringDataState!!.state === DataState.STATE.SUCCESS) {
                Toast.makeText(getThis(), R.string.avatar_change_success, Toast.LENGTH_SHORT).show()
                intent.getStringExtra("friendId")?.let { viewModel!!.startFetchRelationData(it) }
            } else {
                Toast.makeText(applicationContext, "失败", Toast.LENGTH_SHORT).show()
            }
        })
        viewModel!!.relationData!!.observe(this, Observer { friendContactDataState ->
            if (friendContactDataState!!.state === DataState.STATE.SUCCESS) {
                remarkText!!.text = friendContactDataState!!.data!!.remark
            }
        })
    }

    override fun onResume() {
        super.onResume()
        if (intent.hasExtra("friendId")) {
            intent.getStringExtra("friendId")?.let { viewModel!!.startFetchRelationData(it) }
        }
    }
}